clang/test/CodeGenSYCL/sycl-device.cpp

```diff
@@ -3,6 +3,78 @@
 
 int bar(int b);
 
+int bar10(int a) { return a + 10; }
+int bar20(int a) { return a + 20; }
+
+class A {
+public:
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1A3fooEv
+  // CHECK-DAG: define spir_func i32 @_Z5bar20i
+  __attribute__((sycl_device)) void foo() { bar20(10); }
+
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1AC1Ev
+  __attribute__((sycl_device))
+  A() { bar10(10); }
+  // CHECK-DAG: define spir_func i32 @_Z5bar10i
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1AD1Ev
+  __attribute__((sycl_device)) ~A() {}
+
+  template <typename T>
+  __attribute__((sycl_device)) void AFoo(T t) {}
+
+  // Templates are emitted when they are instantiated
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1A4AFooIiEEvT_
+  template <>
+  __attribute__((sycl_device)) void AFoo<int>(int t) {}
+};
+
+template <typename T>
+struct B {
+  T data;
+  B(T _data) : data(_data) {}
+
+  __attribute__((sycl_device)) void BFoo(T t) {}
+};
+
+template <>
+struct B<int> {
+  int data;
+  B(int _data) : data(_data) {}
+
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN1BIiE4BFooEi
+  __attribute__((sycl_device)) void BFoo(int t) {}
+};
+
+struct Base {
+  // CHECK-DAG: define linkonce_odr spir_func void @_ZN4Base12BaseWithAttrEv
+  __attribute__((sycl_device)) virtual void BaseWithAttr() { int a = 10; }
```

Comment:
Can we call virtual function on the device?

Reply:
We can't now. The diagnostic which we implemented to diagnose virtual function call will be triggered for every virtual function call, even if it doesn't bring usage of virtual table to the device code.

Reply:
Exactly, so what is the point to emit LLVM IR for the function which can't be called? Shouldn't we emit an error instead?

Reply:
I had an attempt to implement diagnosing the case when sycl_device attribute is applied to a virtual function.
Unfortunately, this diagnostic can't work well.
It seems that in place where clang is adding some explicit attributes to a declaration, we can understand that function is virtual only if it has explicit `virtual` specifier. So, if function is still virtual, but it is not explicitly marked as `virtual`, the diagnostic won't be triggered.

Meanwhile, we have a diagnostic to prevent calling of virtual functions in the device code. This diagnostic is implemented in SemaSYCL https://github.com/intel/llvm/blob/092367be7785a7764c6dc41497f2c74c6e164824/clang/lib/Sema/SemaSYCL.cpp#L226
and it is emitted later than handling of explicit attributes, so there is more information. This diagnostic is emitted for virtual functions which are not marked as `virtual` explicitly.
So, It looks like this:
```
struct Bar {                                       
  virtual void bar1() {}                           
};                                                 
                                                   
struct Bar1 : Bar {                                
  __attribute__((sycl_device)) void bar1() final {} // no diagnostic
};                                                 
                                                   
__attribute__((sycl_device)) void FOO() {          
  Bar1 b;                                          
  b.bar1();  // sycl-device.cpp:65:5: error: SYCL kernel cannot call a virtual function
                 // b.bar1();                                                          
}  
```
I already described it in previous conversations https://github.com/intel/llvm/pull/1295#discussion_r392174119

Reply:
Yep, exactly.  This is useful for the future when we figure out how to better limit the 'virtual call' limitation to only those where the call is actually virtual.  In those cases, you'd want it emitted.

Reply:
This test case is bit different: we have both attributes explicitly added to the declaration: `__attribute__((sycl_device)) virtual`.
Your point is we don't need to add new diagnostics for this case as it might be detected later. Right?

Reply:
My point that it's better to diagnose virtual function presence in device code only if this function is called. Such diagnostic is more accurate.
And, as Erich mentioned, allowing to apply `sycl_device` attribute to virtual functions can be useful if we decide to diagnose only **real** virtual calls (now we diagnose all calls with virtual callees).

Reply:
Did I answer your question? Can we proceed?

Reply:
If I understand it correctly, virtual member functions will be emitted for the device. Right?
I don't understand how we can make this work if function pointers are not supported.
What are we going to do with the vtable?
Do we expect `vtable  == nullptr` or do we expect all devices to support function pointers?

Reply:
> If I understand it correctly, virtual member functions will be emitted for the device. Right?

Yes, you are right.

> I don't understand how we can make this work if function pointers are not supported.
What are we going to do with the vtable?

Technically, if virtual function call was made when user explicitly specified class type, instead of using pointer to the base class, this doesn't involve vtable usage, i.e.:
```
struct Bar {                                       
  virtual void bar1() {}                           
};                                                 
                                                                                                                                                      
__attribute__((sycl_device)) void FOO() {          
  Bar b;                                          
  b.bar();  // won't bring usage of vtable                          
} 
```
 
BTW, we already allow using of classes with virtual functions. It means that if some class has virtual functions and this class somehow used in device code (but without using virtual members), vtable anyway gets into device IR but it's not used. I think nothing is going crazy just because of SPIRV translator is ok with vtable due to function pointers extension support and after that, I suppose, vtable is removed as dead code by back-ends, since it's not used.

> Do we expect vtable == nullptr or do we expect all devices to support function pointers?

None of this. We already deal with vtable **presence** in the device code, and, I think we can allow calls to virtual functions which don't bring vtable **usage** .